r/src/compute.cpp

```diff
@@ -82,4 +82,148 @@ std::shared_ptr<arrow::Table> Table__cast(
   return arrow::Table::Make(schema, std::move(columns), table->num_rows());
 }
 
+// [[arrow::export]]
+std::shared_ptr<arrow::Array> Array__Take(
+    const std::shared_ptr<arrow::Array>& values,
+    const std::shared_ptr<arrow::Array>& indices) {
+
+  std::shared_ptr<arrow::Array> out;
+  arrow::compute::FunctionContext context;
+  arrow::compute::TakeOptions options;
+  STOP_IF_NOT_OK(arrow::compute::Take(&context, *values, *indices, options, &out));
+  return out;
+}
+
+// [[arrow::export]]
+std::shared_ptr<arrow::RecordBatch> RecordBatch__Take(
+    const std::shared_ptr<arrow::RecordBatch>& batch,
+    const std::shared_ptr<arrow::Array>& indices) {
+  int ncols = batch->num_columns();
+  auto nrows = indices->length();
+
+  std::vector<std::shared_ptr<arrow::Array>> columns(ncols);
+
+  for (R_xlen_t j = 0; j < ncols; j++) {
+    columns[j] = Array__Take(batch->column(j), indices);
+  }
+
+  return arrow::RecordBatch::Make(batch->schema(), nrows, columns);
+}
+
+// [[arrow::export]]
+std::shared_ptr<arrow::ChunkedArray> ChunkedArray__Take(
```

Comment:
shouldn't we try to implement this ChunkedArray.Take logic in C++ ? (so python and others also benefit)

Reply:
I would think so, but (1) it was faster for me to prototype here and (2) it wasn't clear to me what the expectations for this function would be, so I figured I'd write something and that could spark some discussion.

IMO I'd prefer to merge this PR with the code where it is and open a followup Jira to relocate this and some other functions to the C++ library.

Reply:
Same.

Reply:
Sure!

Reply:
https://issues.apache.org/jira/browse/ARROW-6784